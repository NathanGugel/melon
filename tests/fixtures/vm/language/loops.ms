
let a = 10;

while (a >= 0)
{
    io.print("Iteration: " .. string.toString(a));
    a = a - 1;
}

while (a > 1)
{
    io.print("Don't print this");
}

let b = 10;
let keepGoing = true;

while (b >= 0 && keepGoing)
{
    io.print("Keep going!");

    if (b == 6)
    {
        keepGoing = false;
    }

    b = b - 1;
}

for (let i in 0..5)
{
   io.print("For iteration: " .. string.toString(i));
}

// Break/Continue

a = 0;
while (true)
{
    if (a >= 10)
    {
        break;
    }

    io.print("Not breaking: " .. string.toString(a));
    a = a + 1;
}
b = 0;

while (true)
{
    a = 0;

    while (true)
    {
        if (a >= 10)
        {
            break;
        }

        io.print("Not breaking inner: " .. string.toString(a));
        a = a + 1;
    }

    if (b >= 3)
    {
        break;
    }

    io.print("Not breaking outer: " .. string.toString(b));
    b = b + 1;
}

a = 0;
while (a < 10)
{
    a = a + 1;

    if (a % 2 == 0)
    {
        continue;
    }

    io.print("Continue passed: " .. string.toString(a));
}

b = 0;
while (b < 5)
{
    b = b + 1;

    if (b % 2 == 1)
    {
        continue;
    }

    io.print("Continue outer passed: " .. string.toString(b));

    a = 0;

    while (a < 10)
    {
        a = a + 1;

        if (a % 2 == 0)
        {
            continue;
        }

        io.print("Continue inner passed: " .. string.toString(a));
    }
}

a = 0;
while (true)
{
    a = a + 1;

    if (a % 2 == 0)
    {
        continue;
    }

    if (a >= 5)
    {
        break;
    }

    io.print("Running: " .. string.toString(a));
}

for (let i1 in 0..10)
{
    if (i1 % 2 == 0)
    {
        continue;
    }

    io.print("For value: " .. string.toString(i1));

    if (i1 == 5)
    {
        break;
    }
}

let arr = [1, 5, 7, 9, 10, 11];

for (let i2 in arr)
{
    if (i2 % 3 == 0)
    {
        continue;
    }

    io.print("For value: " .. string.toString(i2));

    if (i2 == 10)
    {
        break;
    }
}