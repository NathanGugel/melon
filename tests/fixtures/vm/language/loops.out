Iteration: 10
Iteration: 9
Iteration: 8
Iteration: 7
Iteration: 6
Iteration: 5
Iteration: 4
Iteration: 3
Iteration: 2
Iteration: 1
Iteration: 0
Keep going!
Keep going!
Keep going!
Keep going!
Keep going!
For iteration: 0
For iteration: 1
For iteration: 2
For iteration: 3
For iteration: 4
Not breaking: 0
Not breaking: 1
Not breaking: 2
Not breaking: 3
Not breaking: 4
Not breaking: 5
Not breaking: 6
Not breaking: 7
Not breaking: 8
Not breaking: 9
Not breaking inner: 0
Not breaking inner: 1
Not breaking inner: 2
Not breaking inner: 3
Not breaking inner: 4
Not breaking inner: 5
Not breaking inner: 6
Not breaking inner: 7
Not breaking inner: 8
Not breaking inner: 9
Not breaking outer: 0
Not breaking inner: 0
Not breaking inner: 1
Not breaking inner: 2
Not breaking inner: 3
Not breaking inner: 4
Not breaking inner: 5
Not breaking inner: 6
Not breaking inner: 7
Not breaking inner: 8
Not breaking inner: 9
Not breaking outer: 1
Not breaking inner: 0
Not breaking inner: 1
Not breaking inner: 2
Not breaking inner: 3
Not breaking inner: 4
Not breaking inner: 5
Not breaking inner: 6
Not breaking inner: 7
Not breaking inner: 8
Not breaking inner: 9
Not breaking outer: 2
Not breaking inner: 0
Not breaking inner: 1
Not breaking inner: 2
Not breaking inner: 3
Not breaking inner: 4
Not breaking inner: 5
Not breaking inner: 6
Not breaking inner: 7
Not breaking inner: 8
Not breaking inner: 9
Continue passed: 1
Continue passed: 3
Continue passed: 5
Continue passed: 7
Continue passed: 9
Continue outer passed: 2
Continue inner passed: 1
Continue inner passed: 3
Continue inner passed: 5
Continue inner passed: 7
Continue inner passed: 9
Continue outer passed: 4
Continue inner passed: 1
Continue inner passed: 3
Continue inner passed: 5
Continue inner passed: 7
Continue inner passed: 9
Running: 1
Running: 3
For value: 1
For value: 3
For value: 5
For value: 1
For value: 5
For value: 7
For value: 10
